File: 438-find-all-anagrams-in-a-string/find-all-anagrams-in-a-string.cpp
class Solution {
public:
    vector<int> findAnagrams(string s, string p) {
        int i= 0;
        int j= 0;

        int m = s.size();
        int n = p.size();

        vector<int> Pchar(26,0);
        vector<int> Schar(26,0);
        vector<int> ans;

        for(char i : p)
        {
            Pchar[i - 'a']++;
        }

        while ( j < m)
        {
            Schar[s[j] - 'a']++;

            while (j-i+1 > n)
            {
                Schar[s[i] - 'a']--;
                i++;
            }

            if(j-i+1 == n)
            {
                if ( Pchar == Schar) ans.push_back(i);
            }
            j++;
        }
        return ans;
    }
};

File: 438-find-all-anagrams-in-a-string/find-all-anagrams-in-a-string.py
class Solution:
    def findAnagrams(self, s: str, p: str) -> List[int]:
        m = len(s)
        n = len(p)
        ans = []

        pCtr = Counter(p)

        for i in range(0,len(s)-n+1):
            sCtr = Counter(s[i:i+n])
            if pCtr == sCtr:
                ans.append(i)

        return ans

